mitodl/open-discussions
# Widget Editing in Infinite Corridor

#### Pre-Flight checklist


#### What are the relevant tickets?
#3694

#### What's this PR do?
This PR adds basic management support for widgets in Infinite Corridor, currently limited to a single widget type (plaintext) to be extended soon.

#### How should this be manually tested?
1. Navigate to a field page as a field moderator. Click the gear icon, and "Manage Widgets".
2. **Check** that the page URL ends in "/manage/widgets".
3. Make edits to widgets. You should be able to:
    - add new widgets *only "rich" text for now*
    - edit existing widgets
    - delete widgets

   **Note 1:** These changes will not be persisted to the database until you click "Done". I.e., submitting the modal does not make API calls. 

    **Note 2:** There is a drag handle, but it does not do anything yet.
4. Additionally, you should be able to:
    - expand/collapse individual widget contents
    - expand/collapse all widgets
5. Other things to check:
     - Clicking "Cancel" (next to "Done") abandons your changes
     - Clicking "Cancel" or "Done" re-routes you to the field page

#### Any background context you want to provide?
The majority of the new code is in `ol-widgets`. My goal was:
 - `ol-widgets` handles most complex logic and exposes `WidgetsListEditable`, which is basically a form that you can pass handlers and values.
 - `ol-widgets` takes care of a little bit of styling itself, but the majority of styling is handled within Infinite Corridor by passing a few classes to `WidgetsListEditable`
 - API cals are in infinite corridor

#### Screenshots (if appropriate)

**Entering Widget Management:**
<img width="600" alt="Screen Shot 2022-08-19 at 4 43 56 PM" src="https://user-images.githubusercontent.com/9010790/185705095-377c4b52-0cd0-4b09-b378-3136a34fab96.png">

**Widget Management Sidebar:**
<img width="600" alt="Screen Shot 2022-08-19 at 4 28 11 PM" src="https://user-images.githubusercontent.com/9010790/185702672-e012dc47-5a03-44e3-8a6e-9749fc20371c.png">

**Editing an existing widget:**
<img width="600" alt="Screen Shot 2022-08-19 at 4 28 24 PM" src="https://user-images.githubusercontent.com/9010790/185702710-204340e7-103d-4cf6-b244-2a03b4d05626.png">

**Adding a new widget:**
<img width="600" alt="Screen Shot 2022-08-19 at 4 28 33 PM" src="https://user-images.githubusercontent.com/9010790/185702747-f5b8d658-3540-494f-b5ce-50b797af661f.png">

<img width="600" alt="Screen Shot 2022-08-19 at 4 28 36 PM" src="https://user-images.githubusercontent.com/9010790/185702835-4535bd13-83fd-4c8b-b0c8-6b4c7da09a81.png">
